src/appservice/AppService.ts

```diff
@@ -0,0 +1,150 @@
+            // It also allows us to combine constructor/initialize logic
+            // to make the code base much simpler. A small hack to pay for an overall less hacky code base.
+            controller: {
+                onUserQuery: () => {throw new Error("Mjolnir uninitialized")},
+                onEvent: () => {throw new Error("Mjolnir uninitialized")},
+            },
+            suppressEcho: false,
+        });
+        await bridge.initalise();
+        const accessControlListId = await bridge.getBot().getClient().resolveRoom(config.accessControlList);
+        const accessControl = await AccessControl.setupAccessControl(accessControlListId, bridge);
+        const mjolnirManager = await MjolnirManager.makeMjolnirManager(dataStore, bridge, accessControl);
+        const appService = new MjolnirAppService(
+            config,
+            bridge,
+            mjolnirManager,
+            accessControl,
+            dataStore
+        );
+        bridge.opts.controller = {
+            onUserQuery: appService.onUserQuery.bind(appService),
+            onEvent: appService.onEvent.bind(appService),
+        };
+        return appService;
+    }
+
+    /**
+     * Start the appservice for the end user with the appropriate settings from their config and registration file.
+     * @param port The port to make the appservice listen for transactions from the homeserver on (usually sourced from the cli).
+     * @param config The parsed configuration file.
+     * @param registrationFilePath A path to their homeserver registration file.
+     */
+     public static async run(port: number, config: IConfig, registrationFilePath: string) {
+        const dataStore = new PgDataStore(config.db.connectionString);
+        await dataStore.init();
+        const service = await MjolnirAppService.makeMjolnirAppService(config, dataStore, registrationFilePath);
+        // Can't stress how important it is that listen happens last.
+        await service.start(port);
+    }
+
+    public onUserQuery (queriedUser: MatrixUser) {
+        return {}; // auto-provision users with no additonal data
+    }
+
+    /**
+     * Handle an individual event pushed by the homeserver to us.
+     * This function is async (and anything downstream would be anyway), which does mean that events can be processed out of order.
+     * Not a huge problem for us, but is something to be aware of.
+     * @param request A matrix-appservice-bridge request encapsulating a Matrix event.
+     * @param context Additional context for the Matrix event.
+     */
+    public async onEvent(request: Request<WeakEvent>, context: BridgeContext) {
+        const mxEvent = request.getData();
+        if ('m.room.member' === mxEvent.type) {
+            if ('invite' === mxEvent.content['membership'] && mxEvent.state_key === this.bridge.botUserId) {
+                await this.mjolnirManager.provisionNewMjolnir(mxEvent.sender);
```

Comment:
So we provision a new Mjölnir for each room? I kinda assume that we could use the same Mjölnir for several rooms.

Thinking about it a bit more, I think that a Mjölnir per room makes more sense, actually, since it will be easier to somehow hand out the Mjölnir to another user. Is this choice documented somewhere?

Also, just to be sure: if you invite a Mjölnir `@mjolnir:example.org`, since you provision a new Mjölnir, it will be a new instance e.g. `@_mjolnir-XXXX:example.org` joining, right?

Reply:
Nope, i should document what this listener is...

Reply:
appservices have a specific bot designated by `sender_localpart` in the registration file that can be used for admin functions of the appservice.
If you invite the appservice bot to a room it will start the provisioning process and give you a mjolnir.